### PR TITLE
Water && Fire @ THE BESTIES: Sandy, Ida, Beatrice and Hanh Solo

### PR DESCRIPTION
**Pull request by seattlefurby17**
# Assignment Submission: bEtsy

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions. These should be answered by all members of your team, not by a single teammate.

## Reflection

Prompt | Response
--- | ---
Each team member: what is one thing you were primarily responsible for that you're proud of? | 
Hanh: Testings
Ida: proud of contributing to the order fulfillment page, specifically implementing total_revenue and total_num methods
Sandy: adding bootstrap
Beatrice: cart methods
Each team member: what is one thing you were primarily responsible for that you would like targeted feedback on? | 
Hanh: custom methods
Ida: order fulfillment
Sandy: testing for order items theme* 
Beatrice: orders controller
How did your team break up the work to be done? | with Trello cards and feedback templates to assign tasks 
How did your team utilize git to collaborate? | git branch, git merge and PR
What did your group do to try to keep your code DRY while many people collaborated on it? | private methods for share attributes, before_action for controllers, fixtures for tests, and form partials. 
What was a technical challenge that you faced as a group? | validations without !, cookies issues and git out of sync 
What was a team/personal challenge that you faced as a group? | loss time due to personal issues.
What was your application's ERD? (upload this to Google Drive, change the share settings to viewable by everyone with a link, and include a link) | https://drive.google.com/file/d/1ksRVbvnkKLCVpVXj7uuW-Nd0klPX2Igx/view
What is your Trello URL? | https://trello.com/b/qLAW6qrJ/petsy
What is the Heroku URL of your deployed application? | https://petsy-pals.herokuapp.com/

### CODE REVIEW

File: app/assets/stylesheets/application.scss [hunk 1]
```diff
@@ -0,0 +1,10356 @@
...
+@import "bootstrap";
+
+@import "**/*";
+
+/*!
+ * Bootstrap v4.5.3 (https://getbootstrap.com/)
+ * Copyright 2011-2020 The Bootstrap Authors
+ * Copyright 2011-2020 Twitter, Inc.
+ * Licensed under MIT (https://github.com/twbs/bootstrap/blob/main/LICENSE)
+ */
```
Comment (kaidamasaki) [inline app/assets/stylesheets/application.scss:25]: You shouldn't have needed to copy/paste Boostrap into here.  It was already imported on line 16.

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
+class ApplicationController < ActionController::Base
+  # before_action :require_login
```
Comment (kaidamasaki) [inline app/controllers/application_controller.rb:2]: You should try to remember to clean up commented out code before pushing in the future.

File: app/controllers/order_items_controller.rb [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+      flash[:error] = "Invalid quantity to add to cart given"
+      redirect_to products_path
+      return
+    end
+
+    if @product.nil?
+      flash[:error] = "Product to add to cart not found"
+      redirect_to products_path
+      return
+    end
```
Comment (kaidamasaki) [inline app/controllers/order_items_controller.rb:15]: This logic should probably have lived in a controller filter, or at least a helper method (since you likely want to share this logic with `edit_quantity` and `destroy`).

File: app/controllers/orders_controller.rb [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+      @order.errors.each do |type, err|
+        flash.now[type] = "Something went wrong: #{type}: #{err}"
+      end
+      render :check_out
+    end
+  end
+
+  private
+
+  def find_order
```
Comment (kaidamasaki) [inline app/controllers/orders_controller.rb:65]: This just being named `find_order` makes me expect it will do it by id.  I would probably have named it something like: `find_order_for_shopper`

File: app/controllers/products_controller.rb [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+      redirect_to product_path(@product)
+      return
+    end
+  end
+
+  private
+
+  def find_product
+    @product = Product.find_by(id: params[:id])
+  end
```
Comment (kaidamasaki) [inline app/controllers/products_controller.rb:86]: The `flash` and redirect behavior you rolled one off above should probably have been moved into here:

```suggestion
  def find_product
    @product = Product.find_by(id: params[:id])
    
    if @product.nil?
      flash[:error] = "Cannot find product"
      redirect_to products_path
      return
    end
  end
```

File: app/models/order.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  has_many :products, through: :order_items
+
+  # validates_associated :order_items # one way street, infinite loop otherwise
+  validates :name, format: { :with => /\A[a-zA-Z]+\z/, :message => "Only letters allowed" }, :allow_nil => true
+  validates :email, format: { with: /\A([^@\s]+)@((?:[-a-z0-9]+\.)+[a-z]{2,})\z/i }, :allow_nil => true
+  validates :card_number, numericality: { only_integer: true, :message => "Invalid card number" }, :allow_nil => true
+  validates :expiration_month, numericality: { only_integer: true, greater_than: 0, less_than: 13 }, :allow_nil => true
+  validate  :expiration_year_cannot_be_in_the_past
+  validates :security_code, numericality: { only_integer: true}, :allow_nil => true
+  validates :zip_code,  numericality: { only_integer: true}, :allow_nil => true
```
Comment (kaidamasaki) [inline app/models/order.rb:12]: Woah, these are intense validations.  Nice!

File: app/views/layouts/application.html.erb [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+    <style>
+       /*h1 {*/
+       /*    font-family: 'Indie Flower', cursive;*/
+       /*}*/
+       /*p {*/
+       /*    font-family: 'Indie Flower', cursive;*/
+
+       /*}*/
+    </style>
```
Comment (kaidamasaki) [inline app/views/layouts/application.html.erb:25]: This empty style block is kind of weird.

Comment (kaidamasaki) [inline app/views/layouts/application.html.erb:25]: This style block is a little funky.  I assume it's leftovers from debugging.

File: app/views/orders/cart.html.erb [hunk 8]
```diff
@@ -0,0 +1,31 @@
+<h2 class = "spotlight-other"><strong>Shopping Cart:</strong></h2>
+
+<% if @order.order_items.empty? %>
+  <h3 class ="spotlight-padding">Your cart is empty!</h3>
+  <h3 class = "spotlight-padding"><%= link_to "Keep Shopping?", products_path %></h3>
```
Comment (kaidamasaki) [inline app/views/orders/cart.html.erb:6]: You shouldn't use `h3`s for styling like this.  You should be able to take all of the headings from a page and make a table of contents.

It's not something that's often done by sighted people but is really important if someone is trying to access your site with a screen reader.

File: app/controllers/order_items_controller.rb [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    if @product.retired
+      flash[:error] = "That product is retired!"
+      redirect_to products_path
+      return
+    end
+    # Check if the item exists
+    if @order_item = OrderItem.find_by(product_id: @product.id, order_id: @shopper)
+      # Add to the existing quantity instead of creating a new item
+      @order_item.add_quantity(params["quantity"].to_i)
+      @order_item.save
```
Comment (kaidamasaki) [inline app/controllers/order_items_controller.rb:27]:
```suggestion
      @order_item.save!
```

File: app/models/merchant.rb [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    end
+    return orders
+  end
+
+  def order_belongs_to_merchant?(order)
+    return false if order.nil?
+
+    order.products.each do |product|
+      return true if self.products.include?(product)
+    end
```
Comment (kaidamasaki) [inline app/models/merchant.rb:54]: I think this loop always returns true.

I suspect there should have been a comparison to the merchant in here somewhere.

File: app/models/order.rb [hunk 11]
```diff
@@ -0,0 +1,33 @@
+class Order < ApplicationRecord
+  has_many :order_items
+  has_many :products, through: :order_items
+
+  # validates_associated :order_items # one way street, infinite loop otherwise
+  validates :name, format: { :with => /\A[a-zA-Z]+\z/, :message => "Only letters allowed" }, :allow_nil => true
+  validates :email, format: { with: /\A([^@\s]+)@((?:[-a-z0-9]+\.)+[a-z]{2,})\z/i }, :allow_nil => true
```
Comment (kaidamasaki) [inline app/models/order.rb:7]: Nice email regex.  I assume you grabbed it form somewhere.

Fun fact, there's an input type for email that will do client side validation for you: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/email

(You still need server side validation because of stuff like Postman though.)

File: app/models/product.rb [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+  belongs_to :merchant
+  has_many :order_items
+  has_many :orders, through: :order_items
+
+  validates :name, presence: true, uniqueness: true
+  validates :price, presence: true, numericality: { greater_than: 0 }
+
+  def self.spotlight
+    spotlight_for_all_products = Product.all.sample
+    return spotlight_for_all_products
```
Comment (kaidamasaki) [inline app/models/product.rb:11]: You don't need to assign the result to a variable here since you immediately return it:

```suggestion
    return Product.all.sample
```

**Review by kaidamasaki — COMMENTED** [review]
Great job!  Your project made me super nostalgic which I think was exactly what you were setting out for.

While there were few rough edges you worked together great as a team and overcame a number of personal difficulties together, congratulations on making a working online store!

This project was intentionally a huge lift and you rose to the challenge admirably!

Also, don't take the rubric too seriously it's mostly to just make sure we look at everything important.  (There's a reason we don't give grades depending on the number of check marks...)

I listed you as yellow mostly because of missing reviews, but a yellow on bEtsy is still an impressive accomplishment!

# bEtsy



## Functional Requirements: Manual Testing

Workflow | yes / no
---   | ---
**Before logging in** |
Browse all products, by category, by merchant | Categories not implemented.
Leave a review | Reviews not implemented.
Verify unable to create a new product | ✔️
**After logging in** | |
Create a category | Categories not implemented.
Create a product in that category with stock 10 | ✔️
Add the product you created to your cart | ✔️
Add it again (should update quantity) | ✔️
Verify unable to increase quantity beyond stock | Not implemented.
Add another merchant's product | ✔️
Check out | ✔️
Check that stock was reduced | Not implemented.
Change order-item's status on dashboard | ✔️
Verify unable to leave a review for your own product | Reviews not implemented.
Verify unable to edit another merchant's product by manually editing URL | 
Verify unable to see another merchant's dashboard by manually editing URL | ✔️

## Major Learning Goals/Code Review

Criteria | yes / no
---  | ---
90% reported coverage for all controller and model classes using SimpleCov | ✔️ 98.72%!  :tada:
**Routes** |
No un-needed routes generated (check `reviews`) | ✔️
Routes not overly-nested (check products and merchants) | ✔️
Merchant dashboard and cart page use a non-parameterized routes (should pull merchant or cart ID from session) | ✔️
**Controllers** |
Controller-filter to require login is applied to all merchant-specific actions (update/add item, add category, view merchant dashboard, etc.) - filter method is not duplicated across multiple files | ✔️
Helper methods or filters to find logged-in user, cart, product, etc | ✔️
No excessive business logic | ✔️
**Business logic that ought to live in the model** |
Add / remove / update product on order | Not implemented.
Checkout -> decrease inventory | Not implemented.
Merchant's total revenue | ✔️
Find all orders for this merchant (instance method on `Merchant`) | ✔️
**Selected Model Tests** |
Add item to cart:<br>  - Can add a good product<br>  - Can't add a product w/o enough stock<br>  - Can't add a retired product<br>  - Can't add to an order that's not in cart mode<br>  - Logic specific to this implementation | Some edge cases missing.
Get orders for this merchant:<br>  - Includes all orders from this merchant<br>  - Doesn't include orders from another merchant<br>  - Orders are not included more than once<br>  - Does something reasonable when there are no orders for this merchant | Some edge cases missing.
**Selected Controller Tests** |
Add item to cart:<br>  - Empty cart (should be created)<br>  - Cart already exists (should add to same order)<br>  - Product already in cart (should update quantity)<br>  - Bad product ID, product is retired, quantity too high, or something like that (error) | Some edge cases missing.
Leave a review:<br>  - Works when not logged in<br>  - Works when logged in as someone other than the product's merchant<br>  - Doesn't work if logged in as this product's merchant<br>  - Doesn't work if validations fail | Reviews not implemented.



## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅





## Only the person who submitted the PR will get an email about this feedback. Please let the rest of your team know about it.
